consensus/state_test.go

```diff
@@ -1028,6 +1030,31 @@ func TestSetValidBlockOnDelayedPrevote(t *testing.T) {
 	assert.True(t, rs.ValidRound == round)
 }
 
+// regression for #2518
+func TestNoHearbeatWhenNotValidator(t *testing.T) {
+	cs, _ := randConsensusState(4)
+	cs.Validators = types.NewValidatorSet(nil) // make sure we are not in the validator set
+
+	cs.evsw.AddListenerForEvent("testing", types.EventProposalHeartbeat,
+		func(data tmevents.EventData) {
+			t.Errorf("Should not have broadcasted heartbeat")
+		})
+	go cs.proposalHeartbeat(10, 1)
```

Comment:
do we need to run this in a goroutine?

Reply:
I have no idea. I did this for two reasons:
1) calling `proposalHearbeat(..)` blocks because of the for-loop in the implementation and I have not enough golang experience to have another solution at hand
2) it's the same as it is called in the implementation